rust/arrow/src/record_batch.rs

```diff
@@ -34,36 +35,61 @@ pub struct RecordBatch {
 }
 
 impl RecordBatch {
-    pub fn new(schema: Arc<Schema>, columns: Vec<ArrayRef>) -> Self {
-        // assert that there are some columns
-        assert!(
-            columns.len() > 0,
-            "at least one column must be defined to create a record batch"
-        );
-        // assert that all columns have the same row count
+    /// Creates a `RecordBatch` from a schema and columns
+    ///
+    /// Expects the following:
+    ///  * the vec of columns to not be empty
+    ///  * the schema and column data types to have equal lengths and match
+    ///  * each array in columns to have the same length
+    pub fn new(schema: Arc<Schema>, columns: Vec<ArrayRef>) -> Result<Self> {
+        // check that there are some columns
+        if columns.is_empty() {
+            return Err(ArrowError::ParseError(
```

Comment:
nit: not sure `ParseError` is appropriate here. Do we have other candidate such as `InvalidArgumentError`?

Reply:
`InvalidArgumentError` sounds good, I was avoiding adding another error option